stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,269 @@ public func gradient<T, U, V, R>(
+  func _moved(along direction: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _tangentVector(from cotangent: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  /// The underlying base value.
+  var _base: Any { get }
+
+  /// Returns the underlying value unboxed to the given type, if possible.
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+
+  /// Returns true if the underlying value has type `AnyDerivative.OpaqueZero`.
+  func _isOpaqueZero() -> Bool
+}
+
+internal struct _ConcreteDerivativeBox<T> : _AnyDerivativeBox
+  where T : Differentiable, T.TangentVector == T,
+        T.AllDifferentiableVariables == T,
+        // NOTE: The requirement below should be defined on `Differentiable`.
+        // But it causes a crash due to generic signature minimization bug.
+        T.CotangentVector == T.CotangentVector.AllDifferentiableVariables
+{
+  /// The underlying base value.
+  var _baseDerivative: T
+
+  public init(_ base: T) {
+    self._baseDerivative = base
+  }
+
+  /// The underlying base value, type-erased to `Any`.
+  var _base: Any {
+    return _baseDerivative
+  }
+
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+  {
+    return (self as _AnyDerivativeBox as? _ConcreteDerivativeBox<U>)?
+      ._baseDerivative
+  }
+
+  func _isOpaqueZero() -> Bool {
+    return T.self == AnyDerivative._OpaqueZero.self
+  }
+
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool {
+    // 0 == 0 => true
+    if _isOpaqueZero() && other._isOpaqueZero() {
+      return true
+    }
+    if _isOpaqueZero() {
+      return type(of: other)._zero._isEqual(to: other)
+    }
+    if other._isOpaqueZero() {
+      return _ConcreteDerivativeBox<T>(T.zero)._isEqual(to: self)
+    }
+    guard let otherBase = other._unboxed(to: T.self) else {
+      fatalError()
+    }
+    return _baseDerivative == otherBase
+  }
+
+  func _isNotEqual(to other: _AnyDerivativeBox) -> Bool {
+    // 0 != 0 => false
+    if _isOpaqueZero() && other._isOpaqueZero() {
+      return false
+    }
+    if _isOpaqueZero() {
+      return type(of: other)._zero._isNotEqual(to: other)
+    }
+    if other._isOpaqueZero() {
+      return _ConcreteDerivativeBox<T>(T.zero)._isNotEqual(to: self)
+    }
+    guard let otherBase = other._unboxed(to: T.self) else {
+      fatalError()
```

Comment:
Should return `false`. See `AnyHashable.==` and `AnyHashable.!=`'s behavior.

All you need in this function is
```swift
return _base != other._unboxed(to: T.self)
```

Same for `_isEqual(to:)`.

Reply:
Done.